File: python/Iterator/menu.py
from collections import Iterable, deque
from typing import Deque, List
from dataclasses import dataclass
from abc import ABC, abstractmethod


@dataclass
class MenuItem:
    name: str
    description: str
    vegetarian: bool
    price: float


class Menu(ABC):
    @abstractmethod
    def createIterator(self) -> Iterable[MenuItem]:
        pass


class PancakeHouseMenu(Menu):
    def __init__(self) -> None:
        self.menu_items: Deque[MenuItem] = deque()
        self.add_item(
            "K&P Pancake Breakfast", "Pancakes with scrambled eggs,toast", True, 2.99
        )
        self.add_item(
            "Regular Pancake Breakfast", "Pancakes with fried eggs,sausage", False, 2.99
        )
        self.add_item(
            "Blueberry Pancake", "Pancakes made with fresh blueberries", True, 3.49
        )
        self.add_item(
            "Waffles",
            "Waffles, with your choice of blueberries or strawberries",
            True,
            3.59,
        )

    def createIterator(self) -> Iterable[MenuItem]:
        return self.menu_items

    def add_item(self, name: str, description: str, vegetarian: bool, price: float):
        menu_item = MenuItem(name, description, vegetarian, price)
        self.menu_items.append(menu_item)


class DinerMenu(Menu):
    def __init__(self) -> None:
        self.menu_items: List[MenuItem] = list()
        self.add_item(
            "Vegetarian BLT",
            "(Fakin')Bacon with lettuce & tomato on whole wheat",
            True,
            2.99,
        )
        self.add_item("BLT", "Bacon with lettuce & tomato on whole wheat", False, 2.99)
        self.add_item(
            "Soup of the day",
            "Soup of the day, with a side of potato salad",
            False,
            3.29,
        )
        self.add_item(
            "Hotdog",
            "A hot dog, with saurkraut, relish, onions, topped with cheese",
            False,
            3.05,
        )

    def createIterator(self) -> Iterable[MenuItem]:
        return self.menu_items

    def add_item(self, name: str, description: str, vegetarian: bool, price: float):
        menu_item = MenuItem(name, description, vegetarian, price)
        self.menu_items.append(menu_item)

File: c/Iterator/menu.c
#include "menu.h"

static void _addItem(PancakeHouseMenu *self, str name, str description, bool vegetarian, double price);
static Iterator_MenuItem PancakeHouseMenu_createIterator(PancakeHouseMenu *self);
static Iterator_MenuItem DinerMenu_createIterator(DinerMenu *self);

void PancakeHouseMenuNew(PancakeHouseMenu *self)
{
    self->createIterator = PancakeHouseMenu_createIterator;
    self->menu_item_length = 0;
    _addItem(self, "K&P Pancake Breakfast", "Pancakes with scrambled eggs,toast", true, 2.99);
    _addItem(self, "Regular Pancake Breakfast", "Pancakes with fried eggs,sausage", false, 2.99);
    _addItem(self, "Blueberry Pancake", "Pancakes made with fresh blueberries", true, 3.49);
    _addItem(self, "Waffles", "Waffles, with your choice of blueberries or strawberries", true, 3.59);
}
static void _addItem(PancakeHouseMenu *self, str name, str description, bool vegetarian, double price)
{
    if (self->menu_item_length >= MAX_MENU_ITEM_NUM)
    {
        printf("Can't add more menu item into this menu!");
    }
    self->menu_items[self->menu_item_length].name = name;
    self->menu_items[self->menu_item_length].description = description;
    self->menu_items[self->menu_item_length].vegetarian = vegetarian;
    self->menu_items[self->menu_item_length].price = price;
    MenuItem item = self->menu_items[self->menu_item_length];
    self->menu_item_length++;
}
static Iterator_MenuItem PancakeHouseMenu_createIterator(PancakeHouseMenu *self)
{
    Iterator_MenuItem_New(&self->iter, self->menu_items, self->menu_item_length);
    return self->iter;
}

void DinerMenuNew(DinerMenu *self)
{
    self->createIterator = DinerMenu_createIterator;
    MenuItemArray_init(&self->menu_items);
    MenuItemArray_addItem(&self->menu_items, "Vegetarian BLT", "(Fakin') Bacon with lettuce & tomato on whole wheat", true, 2.99);
    MenuItemArray_addItem(&self->menu_items, "BLT", "Bacon with lettuce & tomato on whole wheat", false, 2.99);
    MenuItemArray_addItem(&self->menu_items, "Soup of the day", "Soup of the day, with a side of potato salad", false, 3.99);
    MenuItemArray_addItem(&self->menu_items, "Hotdog", "A hot dog, with saurkraut, relish, onions, topped with cheese", false, 3.05);
}
static Iterator_MenuItem DinerMenu_createIterator(DinerMenu *self)
{
    Iterator_MenuItem_New(&self->iter, self->menu_items.item, self->menu_items.idx);
    return self->iter;
}